R/package.R

```diff
@@ -0,0 +1,157 @@
+##' Updates a dust model in a package. The user-provided code is
+##' assumed to the in `inst/dust` as a series of C++ files; a file
+##' `inst/dust/model.cpp` will be transformed into a file
+##' `src/model.cpp`.
+##'
+##' If your code provides a class `model` then dust will create C++
+##' functions such as `dust_model_alloc` - if your code also includes
+##' names such as this, compilation will fail due to duplicate
+##' symbols.
+##'
+##' We add "Rcpp attributes" to the created functions, and will run
+##' \code{\link{compileAttributes}} on them once the generated code
+##' has been created.
+##'
+##' @param path Path to the package
+##'
+##' @title Create dust model in package
+##' @return Nothing, this function is called for its side effects
+##' @export
+##' @author Rich Fitzjohn
+dust_package <- function(path) {
+  ## 1. check that the package is legit
+  root <- package_validate(path)
+  path_dust <- file.path(root, "inst/dust")
+  path_src <- file.path(root, "src")
+  path_r <- file.path(root, "R")
+
+  ## 2. find target model files
+
+  ## TODO: support alternative location, perhaps not in package - that
+  ## will suit odin.
+  files <- dir(path_dust, pattern = "\\.cpp$")
+  if (length(files) == 0L) {
+    stop(sprintf("No dust files found in '%s/inst/dust'", root))
+  }
+
+  ## 3. identify destination src files, validate signature
+  package_validate_destination(root, files)
+
+  ## 4. generate code
+  template_r <- read_lines(dust_file("template/dust.R.template"))
+  template_cpp <- read_lines(dust_file("template/dust.cpp"))
+  data <- lapply(file.path(path_dust, files), package_generate,
+                 template_cpp, template_r)
+
+  dir.create(path_src, FALSE, TRUE)
+  dir.create(path_r, FALSE, TRUE)
+  for (d in data) {
+    writeLines(d$cpp, file.path(path_src, d$filename))
+  }
+
+  code_r <- c("## Generated by dust - do not edit",
```

Comment:
It might be nice to add the dust version in here

Reply:
done in i25-rng-interface